### PR TITLE
 feat(823): [6] release and tag trigger

### PR DESCRIPTION
**Pull request by tk3fftk**
## Context
Our users want to start jobs triggered release event same like ~pr or ~commit.

## Objective
- https://github.com/screwdriver-cd/screwdriver/commit/c5fc1ff05d75346f534ba9dc024c359f77892bad : scm token is obtained twice. It's wasteful. And do a little refactor of the test.
- https://github.com/screwdriver-cd/screwdriver/commit/4baa8713e3ef73da6374c851e57346913f48e48d : 
    - enable to parse `~release` and `~tag` with `determineStartFrom()`
    - `sha` is empty when the event is `release` or `tag` so this PR provides to a function which gets `sha` after `parseHook()` with `scm.getCommitRefSha()`

It works locally like this:
image

## References
- Related Issue: https://github.com/screwdriver-cd/screwdriver/issues/823
- Blocked by:
    - https://github.com/screwdriver-cd/data-schema/pull/322
    - https://github.com/screwdriver-cd/scm-router/pull/17
    - https://github.com/screwdriver-cd/scm-github/pull/125/files
    - https://github.com/screwdriver-cd/models/pull/342

### CODE REVIEW

File: plugins/webhooks/index.js [hunk 1]
```diff
@@ -130,10 +161,10 @@ function hasTriggeredJob(pipeline, startFrom) {
  * @param   {PipelineFactory}   pipelineFactory The pipeline factory to get the branch list from
  * @param   {Object}            scmConfig       Has the token and scmUri to get branches
  * @param   {String}            branch          The branch which is committed
- * @param   {String}            type            Triggered event type ('pr' or 'commit')
+ * @param   {String}            type            Triggered GitHub event type ('pr' or 'repo')
```
Comment (tkyi) [inline plugins/webhooks/index.js]: need to add `action` to jsdocs

File: plugins/webhooks/index.js [hunk 2]
```diff
@@ -413,67 +447,58 @@ function pullRequestEvent(pluginOptions, request, reply, parsed) {
...
-            });
-        })
-        .then((scmUri) => {
-            scmConfig.scmUri = scmUri;
+    return pipelineFactory.scm.parseUrl({
+        checkoutUrl: fullCheckoutUrl,
+        token,
+        scmContext
+    }).then((scmUri) => {
+        scmConfig.scmUri = scmUri;
```
Comment (tkyi) [inline plugins/webhooks/index.js]: do you need to pass `token` into `scmConfig` as well?
https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/webhooks/index.js#L409

File: plugins/webhooks/index.js [hunk 3]
```diff
@@ -4,9 +4,40 @@ const boom = require('boom');
...
+ */
+function determineStartFrom(action, type, targetBranch, pipelineBranch) {
+    let startFrom;
+
+    if (type && type === 'pr') {
+        startFrom = '~pr';
+    } else {
+        switch (action) {
+        case 'release':
+            startFrom = '~release';
```
Comment (minz1027) [inline plugins/webhooks/index.js]: how are we going to handle external pipeline release trigger? e.g.`sd@123:~release`

Comment (minz1027) [reply plugins/webhooks/index.js]: but can probably implement this in second pass ^^

File: plugins/webhooks/index.js (hunk 1 above)
Comment (tk3fftk) [reply plugins/webhooks/index.js]: added 👍

File: plugins/webhooks/index.js (hunk 2 above)
Comment (tk3fftk) [reply plugins/webhooks/index.js]: It's needed... Passed.

File: plugins/webhooks/index.js (hunk 3 above)
Comment (tk3fftk) [reply plugins/webhooks/index.js]: I couldn't thought about it... Can we implement in second pass?

**Comment by tkyi** [comment]
Need to pull in new major version of `screwdriver-scm-github` in `package.json` as well: should be `^9.0.0`

File: screwdriver.yaml [hunk 4]
```diff
@@ -3,6 +3,8 @@ shared:
 
 jobs:
     main:
+        annotations:
+            screwdriver.cd/ram: HIGH
```
Comment (tkyi) [inline screwdriver.yaml]: try `TURBO`?

Comment (tk3fftk) [reply screwdriver.yaml]: I try to add `NODE_OPTIONS`. I think `HIGH` is enough.

**Comment by tk3fftk** [comment]
Test is passed locally...
```
  637 passing (4m)


=============================== Coverage summary ===============================
Statements   : 99.86% ( 2908/2912 )
Branches     : 97.85% ( 866/885 )
Functions    : 99.88% ( 804/805 )
Lines        : 99.89% ( 2758/2761 )
================================================================================
```

File: plugins/webhooks/index.js [hunk 5]
```diff
@@ -4,9 +4,40 @@ const boom = require('boom');
...
 
+/**
+ * Determine "startFrom" with type, action and branches
+ * @param {String} action          SCM webhook action type
+ * @param {String} type            Triggered SCM event type ('pr' or 'repo')
+ * @param {String} targetBranch    The branch against which commit is pushed
+ * @param {String} pipelineBranch  The pipeline branch
+ * @returns {String}               startFrom
+ */
+function determineStartFrom(action, type, targetBranch, pipelineBranch) {
```
Comment (wdstar) [inline plugins/webhooks/index.js]: We should write unit test for this function. Its test will be useful in the second pass implementation.

File: plugins/webhooks/index.js [hunk 6]
```diff
@@ -640,40 +689,60 @@ exports.register = (server, options, next) => {
...
+                    }
 
                     if (type === 'pr') {
                         // disregard skip ci for pull request events
-                        return pullRequestEvent(pluginOptions, request, reply, parsed);
+                        return pullRequestEvent(pluginOptions, request, reply, parsed, token);
                     }
 
-                    return pushEvent(pluginOptions, request, reply, parsed, skipMessage);
+                    return pushEvent(pluginOptions, request, reply, parsed, skipMessage, token);
```
Comment (wdstar) [inline plugins/webhooks/index.js]: We should use one `if ... else ...`

Comment (tk3fftk) [reply plugins/webhooks/index.js]: It conflicts with `no-else-return` rule of eslint.

File: plugins/webhooks/index.js (hunk 5 above)
Comment (tk3fftk) [reply plugins/webhooks/index.js]: added!

File: plugins/webhooks/index.js (hunk 3 above)
Comment (tk3fftk) [reply plugins/webhooks/index.js]: I created Issue: https://github.com/screwdriver-cd/screwdriver/issues/1534

**Pull request merged**